cou/utils/openstack.py

```diff
@@ -13,125 +13,129 @@
-
-    charm designate-bind workload_version tracks the version of the deb package bind9.
+    Charm designate-bind workload_version tracks the version of the deb package bind9 (rmadison).
+    For charm gnocchi it was used cmadison.
 
     [0] https://releases.openstack.org/
     [1] https://docs.openstack.org/charm-guide/latest/project/charm-delivery.html
 
-    :return: Ordered dictionary containing the version and the possible Openstack release.
+    :return: Ordered dictionary containing the version and the possible OpenStack release.
     :rtype: OrderedDict
     """
-    charms_types = {
-        "ceph": ["ceph-mon", "ceph-fs", "ceph-radosgw", "ceph-osd"],
-        "swift": ["swift-proxy", "swift-storage"],
-        "nova": ["nova-cloud-controller", "nova-compute"],
-        "ovn": ["ovn-dedicated-chassis", "ovn-central"],
-        "neutron": ["neutron-api", "neutron-gateway"],
-        "manila": ["manila-ganesha"],
-        "horizon": ["openstack-dashboard"],
-    }
-    with open("cou/utils/openstack_lookup.csv") as csv_file:
+    with open(CSV_FILE) as csv_file:
         openstack_lookup = OrderedDict()
-        service_column_index = 0
-        minor_column_index = 1
-        micro_column_index = 2
-        version_start_column_index = 3
         csv_reader = csv.reader(csv_file, delimiter=",")
         header = next(csv_reader)
         for row in csv_reader:
             service_dict: defaultdict[str, Any] = defaultdict(OrderedDict)
-            service = row[service_column_index]
-            minor = row[minor_column_index]
-            micro = row[micro_column_index]
-            for column_index in range(version_start_column_index, len(row), 2):
+            service = row[SERVICE_COLUMN_INDEX]
+            for column_index in range(VERSION_START_COLUMN_INDEX, len(row), 2):
                 os_version, _ = header[column_index].split("-")
                 min_version = row[column_index]
                 max_version = row[column_index + 1]
-                service_dict[os_version] = Versions(
+                service_dict[os_version] = VersionRange(
                     Version(min_version),
                     Version(max_version),
-                    True if minor == "TRUE" else False,
-                    True if micro == "TRUE" else False,
                 )
             openstack_lookup[service] = service_dict
-    for charm_type, charms in charms_types.items():
+    for charm_type, charms in CHARM_TYPES.items():
         for charm in charms:
             openstack_lookup[charm] = openstack_lookup[charm_type]
     return openstack_lookup
 
 
-OPENSTACK_LOOKUP = generate_openstack_lookup()
-
+class OpenStackLookup(object):
+    _OPENSTACK_LOOKUP: OrderedDict = OrderedDict()
+
+    def __new__(cls):
+        """Magic method for create the OpenStackLookup object.
+
+        :return: OpenStackLookup object
+        :rtype: OpenStackLookup
+        """
+        if not cls._OPENSTACK_LOOKUP:
```

Comment:
It should more stable to check if value is None, rather than empty dict. I know that the `generate_openstack_lookup` should never return empty dictionary, but still is good practice.
```suggestion
        if cls._OPENSTACK_LOOKUP is None:
```